File: 2021/day05.hs
import Data.List.Split
import Data.Map.Strict (alter, empty)
import Data.Maybe

main = do
  lines' <- map parseLine . lines <$> readFile "input/day05"
  let points = foldl getPoints empty $ filter isOrthogonal lines'
      points' = foldl getPoints points $ filter isDiagonal lines'
  putStrLn $ "Part 1: " ++ show (countOverlaps points)
  putStrLn $ "Part 2: " ++ show (countOverlaps points')

parseLine = map (map read . splitOn ",") . splitOn " -> "

getPoints acc [[x, y], [x', y']]
  | x == x' && y == y' = acc'
  | x == x' = getPoints acc' [[x, y''], [x', y']]
  | y == y' = getPoints acc' [[x'', y], [x', y']]
  | otherwise = getPoints acc' [[x'', y''], [x', y']]
  where
    acc' = alter (\c -> Just $ fromMaybe 0 c + 1) (x, y) acc
    x'' = x + if x > x' then -1 else 1
    y'' = y + if y > y' then -1 else 1

isOrthogonal [[x, y], [x', y']] = x == x' || y == y'

isDiagonal [[x, y], [x', y']] = abs (x - x') == abs (y - y')

countOverlaps = foldl (\acc c -> acc + if c > 1 then 1 else 0) 0

File: 2021/day05.py
from collections import defaultdict

def draw(canvas, lines):
    for [x0, y0], [x1, y1] in lines:
        xi = 1 if x1 > x0 else (-1 if x0 > x1 else 0)
        yi = 1 if y1 > y0 else (-1 if y0 > y1 else 0)
        x0, y0 = x0 - xi, y0 - yi

        while (x0, y0) != (x1, y1):
            x0, y0 = x0 + xi, y0 + yi
            canvas[(x0, y0)] += 1

    return canvas

parse_line = lambda ln: [list(map(int, c.split(','))) for c in ln.split(' -> ')]
is_orthogonal = lambda ln: ln[0][0] == ln[1][0] or ln[0][1] == ln[1][1]
is_diagonal = lambda ln: abs(ln[0][0] - ln[1][0]) == abs(ln[0][1] - ln[1][1])
count_overlaps = lambda ps: sum(1 for coord in ps if ps[coord] > 1)
lines = list(map(parse_line, open("input/day05", "r").read().strip().split('\n')))
canvas = draw(defaultdict(int), list(filter(is_orthogonal, lines)))

print("Part 1:", count_overlaps(canvas))
print("Part 2:", count_overlaps(draw(canvas, filter(is_diagonal, lines))))
